From ed4a7dda6ac9e64e4bb47659478088c2277b8ae0 Mon Sep 17 00:00:00 2001
From: Koen Kooi <koen@dominion.thruhere.net>
Date: Wed, 8 Jul 2026 16:14:19 +0200
Subject: [PATCH 04/46] Plugwise: appliance ID extraction and stable Domoticz
 IDs

Extract appliance IDs for all appliances and derive stable Domoticz identifiers from the trailing appliance ID bytes.

Signed-off-by: Koen Kooi <koen@dominion.thruhere.net>
Assisted-by: GPT-5.3-Codex

Upstream-Status: Inappropriate [WIP]
---
 hardware/AnnaThermostat.cpp | 68 +++++++++++++++++++++----------------
 1 file changed, 38 insertions(+), 30 deletions(-)

diff --git a/hardware/AnnaThermostat.cpp b/hardware/AnnaThermostat.cpp
index ede09a3ec..6dfc8148c 100644
--- a/hardware/AnnaThermostat.cpp
+++ b/hardware/AnnaThermostat.cpp
@@ -456,6 +456,22 @@ void CAnnaThermostat::GetMeterDetails()
 	while (pAppliance)
 	{
 		TiXmlHandle hAppliance = TiXmlHandle(pAppliance);
+		std::string applianceID = "";
+		unsigned int applianceIDhex;
+
+		pAttribute = pAppliance->FirstAttribute();
+		if (pAttribute != nullptr)
+		{
+			std::string aName = pAttribute->Name();
+			if (aName == "id")
+			{
+				applianceID = pAttribute->Value();
+				applianceIDhex = strtoul(applianceID.substr(28,4).c_str(), NULL, 16);
+				Log(LOG_NORM, "AnnaTherm: Appliance ID: %s, %x", applianceID.c_str(), applianceIDhex);
+			}
+		} else {
+			Log(LOG_ERROR, "AnnaTherm: Cannot find appliance ID");
+		}
 
 		pElem = pAppliance->FirstChildElement("name");
 		if (pElem == nullptr)
@@ -467,15 +483,7 @@ void CAnnaThermostat::GetMeterDetails()
 
 		if ((m_ThermostatID.empty()) && ((ApplianceName == "Anna") || (ApplianceName == "Adam")))
 		{
-			pAttribute = pAppliance->FirstAttribute();
-			if (pAttribute != nullptr)
-			{
-				std::string aName = pAttribute->Name();
-				if (aName == "id")
-				{
-					m_ThermostatID = pAttribute->Value();
-				}
-			}
+			m_ThermostatID = applianceID;
 		}
 
 		//Handle point_log
@@ -503,7 +511,7 @@ void CAnnaThermostat::GetMeterDetails()
 				if (!tmpstr.empty())
 				{
 					float temperature = (float)atof(tmpstr.c_str());
-					SendTempSensor(sAnnaTemperature, 255, temperature, sname);
+					SendTempSensor(applianceIDhex, 255, temperature, sname);
 				}
 			}
 			else if (sname == "illuminance")
@@ -512,7 +520,7 @@ void CAnnaThermostat::GetMeterDetails()
 				if (!tmpstr.empty())
 				{
 					float illuminance = (float)atof(tmpstr.c_str());
-					SendLuxSensor(sAnnaIlluminance, 1, 255, illuminance, sname);
+					SendLuxSensor(applianceIDhex, 1, 255, illuminance, sname);
 				}
 			}
 			else if (sname == "thermostat")
@@ -530,7 +538,7 @@ void CAnnaThermostat::GetMeterDetails()
 				if (!tmpstr.empty())
 				{
 					float temperature = (float)atof(tmpstr.c_str());
-					SendTempSensor(sAnnaIntendedBoilerTemperature, 255, temperature, sname);
+					SendTempSensor(applianceIDhex, 255, temperature, sname);
 				}
 			}
 			else if (sname == "return_water_temperature")
@@ -539,7 +547,7 @@ void CAnnaThermostat::GetMeterDetails()
 				if (!tmpstr.empty())
 				{
 					float temperature = (float)atof(tmpstr.c_str());
-					SendTempSensor(sAnnaReturnWaterTemperature, 255, temperature, sname);
+					SendTempSensor(applianceIDhex, 255, temperature, sname);
 				}
 			}
 			else if (sname == "boiler_temperature")
@@ -548,7 +556,7 @@ void CAnnaThermostat::GetMeterDetails()
 				if (!tmpstr.empty())
 				{
 					float temperature = (float)atof(tmpstr.c_str());
-					SendTempSensor(sAnnaBoilerTemperature, 255, temperature, sname);
+					SendTempSensor(applianceIDhex, 255, temperature, sname);
 				}
 			}
 			else if (sname == "maximum_boiler_temperature")
@@ -557,7 +565,7 @@ void CAnnaThermostat::GetMeterDetails()
 				if (!tmpstr.empty())
 				{
 					float temperature = (float)atof(tmpstr.c_str());
-					SendTempSensor(sAnnaMaxBoilerTemperature, 255, temperature, sname);
+					SendTempSensor(applianceIDhex, 255, temperature, sname);
 				}
 			}
 			else if (sname == "boiler_state")
@@ -567,11 +575,11 @@ void CAnnaThermostat::GetMeterDetails()
 				{
 					if (strcmp(tmpstr.c_str(), "on") == 0)
 					{
-						SendSwitch(sAnnaBoilerState, 1, 255, true, 0, sname, m_Name);
+						SendSwitch(applianceIDhex, 1, 255, true, 0, sname, m_Name);
 					}
 					else
 					{
-						SendSwitch(sAnnaBoilerState, 1, 255, false, 0, sname, m_Name);
+						SendSwitch(applianceIDhex, 1, 255, false, 0, sname, m_Name);
 					}
 				}
 			}
@@ -583,11 +591,11 @@ void CAnnaThermostat::GetMeterDetails()
 					if (strcmp(tmpstr.c_str(), "on") == 0)
 					{
 
-						SendSwitch(sAnnaFlameState, 1, 255, true, 0, sname, m_Name);
+						SendSwitch(applianceIDhex, 1, 255, true, 0, sname, m_Name);
 					}
 					else
 					{
-						SendSwitch(sAnnaFlameState, 1, 255, false, 0, sname, m_Name);
+						SendSwitch(applianceIDhex, 1, 255, false, 0, sname, m_Name);
 					}
 				}
 			}
@@ -616,11 +624,11 @@ void CAnnaThermostat::GetMeterDetails()
 				{
 					if (strcmp(tmpstr.c_str(), "on") == 0)
 					{
-						SendSwitch(sAnnaProximity, 1, 255, true, 0, sname, m_Name);
+						SendSwitch(applianceIDhex, 1, 255, true, 0, sname, m_Name);
 					}
 					else
 					{
-						SendSwitch(sAnnaProximity, 1, 255, false, 0, sname, m_Name);
+						SendSwitch(applianceIDhex, 1, 255, false, 0, sname, m_Name);
 					}
 				}
 			}
@@ -670,7 +678,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float level = (float)atof(tmpstr.c_str()) * 100.F;
-                    SendPercentageSensor(sAnnaModulationLevel, 1, 255, level, sname);
+                    SendPercentageSensor(applianceIDhex, 1, 255, level, sname);
                 }
             }
             else if (sname == "central_heater_water_pressure")
@@ -679,7 +687,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float pressure = (float)atof(tmpstr.c_str());
-                    SendPressureSensor(sAnnaWaterPressure, 1, 255, pressure, sname);
+                    SendPressureSensor(applianceIDhex, 1, 255, pressure, sname);
                 }
             }
             else if (sname == "signal_strength")
@@ -688,7 +696,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float dbm = (float)atof(tmpstr.c_str());
-                    SendCustomSensor(sAnnaSignalStrength, 1, 255, dbm, sname, "dBm");
+                    SendCustomSensor(applianceIDhex, 1, 255, dbm, sname, "dBm");
                 }
             }
             else if (sname == "burner_starts")
@@ -697,7 +705,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float starts = (float)atof(tmpstr.c_str());
-                    SendCustomSensor(sAnnaBurnerStarts, 1, 255, starts, sname, "");
+                    SendCustomSensor(applianceIDhex, 1, 255, starts, sname, "");
                 }
             }
             else if (sname == "failed_burner_starts")
@@ -706,7 +714,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float starts = (float)atof(tmpstr.c_str());
-                    SendCustomSensor(sAnnaFailedBurnerStarts, 1, 255, starts, sname, "");
+                    SendCustomSensor(applianceIDhex, 1, 255, starts, sname, "");
                 }
             }
             else if (sname == "burner_operation_time")
@@ -715,7 +723,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float hour = (float)atof(tmpstr.c_str());
-                    SendCustomSensor(sAnnaBurnerOperationTime, 1, 255, hour, sname, "Hour(s)");
+                    SendCustomSensor(applianceIDhex, 1, 255, hour, sname, "Hour(s)");
                 }
             }
             else if (sname == "domestic_hot_water_burner_operation_time")
@@ -724,7 +732,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float hour = (float)atof(tmpstr.c_str());
-                    SendCustomSensor(sAnnaHotWaterBurnerOperationTime, 1, 255, hour, sname, "Hour(s)");
+                    SendCustomSensor(applianceIDhex, 1, 255, hour, sname, "Hour(s)");
                 }
             }
             else if (sname == "domestic_hot_water_flow")
@@ -733,7 +741,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float flow = (float)atof(tmpstr.c_str());
-                    SendWaterflowSensor(sAnnaHotWaterFlow, 1, 255, flow, sname);
+                    SendWaterflowSensor(applianceIDhex, 1, 255, flow, sname);
                 }
             }
             else if (sname == "domestic_hot_water_temperature")
@@ -742,7 +750,7 @@ void CAnnaThermostat::GetMeterDetails()
                 if (!tmpstr.empty())
                 {
                     float temperature = (float)atof(tmpstr.c_str());
-                    SendTempSensor(sAnnaHotWaterTemperature, 255, temperature, sname);
+                    SendTempSensor(applianceIDhex, 255, temperature, sname);
                 }
             }
         }
